File: tbb/chain/expertseq/winnow.cc
/* winnow: weighted point selection
 *
 * input:
 *   randmat_matrix: an integer matrix, whose values are used as masses
 *   thresh_mask: a boolean matrix, showing which points are eligible for
 *     consideration
 *   nrows, ncols: number of rows and columns
 *   nelts: the number of points to select
 *
 * output:
 *   winnow_points: a vector of (x, y) points
 */
#include <cassert>
#include <cstdio>
#include <cstdlib>
#include <cstring>

#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

extern int is_bench;
extern unsigned char randmat_matrix[20000][20000];
extern unsigned char thresh_mask[20000][20000];
pair<int, int> winnow_points[20000];
static pair<int, pair<int, int> > values[20000];

void winnow(int nrows, int ncols, int nelts) {
  int count = 0;
  for (int i = 0; i < nrows; i++) {
    for (int j = 0; j < ncols; j++) {
      if (is_bench) {
        thresh_mask[i][j] = ((i * j) % (ncols + 1)) == 1;
      }
      if (thresh_mask[i][j]) {
        values[count++] = (make_pair(randmat_matrix[i][j], make_pair(i, j)));
      }
    }
  }

  sort(values, values + count);

  size_t n = count;
  size_t chunk = n / nelts;

  for (int i = 0; i < nelts; i++) {
    int index = i * chunk;
    winnow_points[i] = values[index].second;
  }
}

File: erlang/chain/seq/winnow.erl
%
% winnow: weighted point selection
%
% input:
%   matrix: an integer matrix, whose values are used as masses
%   mask: a boolean matrix showing which points are eligible for
%     consideration
%   nrows, ncols: the number of rows and columns
%   nelts: the number of points to select
%
% output:
%   points: a vector of (x, y) points
%

-module(winnow).
-export([winnow/5]).

get_values_vector(_, _, [], []) -> [];
get_values_vector(Line, Col, [ValuesHead | ValuesTail], [
    MaskHead | MaskTail]) ->
  Rest = get_values_vector(Line, Col + 1, ValuesTail, MaskTail),
  if MaskHead -> [
        {ValuesHead, {Line, Col}} | Rest];
    true -> Rest
  end.

get_values(_, [], []) -> [];
get_values(Line, [MatrixHead | MatrixTail], [MaskHead | MaskTail]) ->
  get_values_vector(Line, 0, MatrixHead, MaskHead) ++ get_values(
    Line + 1, MatrixTail, MaskTail).

drop(L, 0) -> L;
drop([_ | T], I) -> drop (T, I - 1).

get_points(0, _, _) -> [];
get_points(Nelts, [{_, {I, J}} | Tail], Chunk) ->
  [ {I, J} | get_points(Nelts - 1, drop(Tail, Chunk - 1), Chunk)].

winnow(_, _, Matrix, Mask, Nelts) ->
    Values = get_values(0, Matrix, Mask),
    Sorted = lists:sort(Values),
    N = length(Sorted),
    Chunk = N div Nelts,
    Points = get_points(Nelts, Sorted, Chunk),
    Points.
